### ISSUE TITLE
there are two critical missing pieces that will cause headaches during training and inference

### ISSUE THREAD

**Issue by pattanayakpratik** [Shubhamsaboo/awesome-llm-apps#1027]
### ⚠️ The Two Critical Fixes Needed:

1. Pass the `tokenizer` into `SFTTrainer`:

You must explicitly pass `tokenizer=tokenizer` inside the `SFTTrainer`. If you don't, the trainer won't know how to pad the sequences properly or respect the special tokens you just set up in the chat template.
(Highly Recommended: Also add `DataCollatorForSeq2Seq(tokenizer=tokenizer) `to ensure the model learns to mask out the prompt and only calculate loss on the generated answers).

1. Save the `tokenizer` at the end:

You used `model.save_pretrained()`, but you also need to use `tokenizer.save_pretrained()`. If you don't save the tokenizer, your saved model won't remember the` llama-3.1` chat template you applied, which will break inference later.

your code must be 
```
import torch
from unsloth import FastLanguageModel
from datasets import load_dataset
from trl import SFTTrainer, SFTConfig # <-- ADDED SFTConfig (TRL's new standard)
from transformers import DataCollatorForSeq2Seq # <-- ADDED
from unsloth.chat_templates import get_chat_template, standardize_sharegpt

# Load model and tokenizer
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/Llama-3.2-3B-Instruct",
    max_seq_length=2048, 
    load_in_4bit=True,
)

# Add LoRA adapters
model = FastLanguageModel.get_peft_model(
    model, r=16,
    target_modules=[
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj"
    ],
)

# Set up chat template and prepare dataset
tokenizer = get_chat_template(tokenizer, chat_template="llama-3.1")
dataset = load_dataset("mlabonne/FineTome-100k", split="train")
dataset = standardize_sharegpt(dataset)
dataset = dataset.map(
    lambda examples: {
        "text": [
            tokenizer.apply_chat_template(convo, tokenize=False)
            for convo in examples["conversations"]
        ]
    },
    batched=True
)

# Set up trainer
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer, # <-- ADDED: Crucial for proper tokenization and padding
    train_dataset=dataset,
    dataset_text_field="text",
    max_seq_length=2048,
    data_collator=DataCollatorForSeq2Seq(tokenizer=tokenizer), # <-- ADDED: Masks prompts during loss calculation
    args=SFTConfig( # <-- SWAPPED: TrainingArguments to SFTConfig (standard for new TRL versions)
        per_device_train_batch_size=2,
        gradient_accumulation_steps=4,
        warmup_steps=5,
        max_steps=60,
        learning_rate=2e-4,
        fp16=not torch.cuda.is_bf16_supported(),
        bf16=torch.cuda.is_bf16_supported(),
        logging_steps=1,
        output_dir="outputs",
        report_to="none",
    ),
)

# Train the model
trainer.train()

# Save the finetuned model AND tokenizer
model.save_pretrained("finetuned_model")
tokenizer.save_pretrained("finetuned_model") # <-- ADDED: Ensures chat templates load correctly during inference
```

**Comment by Shubhamsaboo** [comment]
solved in #1033
